Feature #1543 concatenation final

jan-petr opened the pull request:
### Linked issue

#1543

jan-petr committed:
#1543 ASLSubtractionAveraging bugfix

jan-petr committed:
#1543 Bug fixes

jan-petr committed:
#1543 Fix JSON saving in wrp_RegisterASL

jan-petr committed:
#1543 Read JSON with NIfTI in original format

jan-petr committed:
#1543 Fix wrp_ResampleASL

jan-petr committed:
#1543 Fix M0 processing

jan-petr committed:
#1543 Move PWI4D merging out of wrp_Quantify

jan-petr committed:
#1543 Fix merging PWI4D with JSONs

jan-petr committed:
#1543 wrp_Quantify use JSONs

jan-petr committed:
#1543 im_ASLSubtractionAveraging accept paths at input

jan-petr committed:
#1543 quant_ASL Minor bug-fix

HenkMutsaerts commented on External/SPMmodified/xASL/xASL_io_Nifti2Im.m:
```diff
@@ -22,7 +22,7 @@
 %              (as explained under INPUT: niftiIn). It does the following.
 %
-%              1. Try to load a NIfTI
+%              1. Try to load a NIfTI and JSON (keeps it in the original format
```
closing bracket ) missing.

Not sure if "keeps it in the original format" is helpful, it could be confusing as if this function also writes. But it only reads/loads.

HenkMutsaerts commented on External/SPMmodified/xASL/xASL_io_ReadNifti.m:
```diff
@@ -90,7 +91,7 @@
 
 if exist(pathJson)
-    json = xASL_bids_parms2BIDS([], xASL_io_ReadJson(pathJson), 0);
+    json = xASL_io_ReadJson(pathJson);
```
Are you sure? I agree that having the BIDS_parms2BIDS separate is perhaps cleaner, but the problem is that we need a "transparent" enforcing of saving JSON always in BIDS and reading always in legacy. Like the nii & nii.gz loading/saving is done without having to write the .gz handling in the code, when reading a JSON we shouldn't have to do xASL_bids_Parms2BIDS in the code.

This is sooo easy to go wrong and can take such a long time to bugfix (when collaborators/users report an issue) that I feel quite strong about this.

I see now below that you perhaps adapted this for the PWI4D merging, but that is a very special case. The NIfTI handling is a much more general thing.

As alternative, we can have a boolean for xASL_io_ReadNifti for converting the loaded JSON to legacy, which is ON by default.

jan-petr replied:
I have described my reasoning in the issue as I thought about it a lot. 

Nifti reading is used on several places (import, processing) and only in processing, we use the Legacy format. Which is a format we are slowly trying to eradicate. So that's why I wanted to avoid doing any conversion and only use explicit conversion for now, as we will be phasing out all legacy features. 

So we can incorporate that, but only with a boolean. Applies of course to all mention below - if we agree to change that then on all occurences.

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m:
```diff
@@ -19,10 +19,14 @@
 %             (xASL_io_Nifti2Im below allows both path and image matrix inputs).
 %             Best if this is a path, so any motion correction can be applied.
-%   PWI4D   - likewise, see explanation below. If this is provided, it will override the PWI4D calculation below (STRING OR 4D MATRIX, OPTIONAL,
+%   PWI4D      - likewise, see explanation below. If this is provided, it will override the PWI4D calculation below (PATH OR 4D MATRIX, OPTIONAL,
```
This is fine but seems slightly less correct. String & matrix are Matlab variable types, a path is a content. So it should be either

PATH or 4D IMAGE

OR

STRING or 4D MATRIX

But I don't really care so you can close this :)

jan-petr replied:
Actually, it was STRING - which wasn't really correct :) So I changed it to CHAR. 

But then realized that 4D MATRIX isn't a type. CHAR would be actually a 1D MATRIX. So PATH vs IMAGE or PATH vs 4D MATRIX seemed best.

No strong opinion, but not STRING.

HenkMutsaerts replied:
OK we can do path vs 4D image?

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m:
```diff
@@ -19,10 +19,14 @@
+%   PWI4D      - likewise, see explanation below. If this is provided, it will override the PWI4D calculation below (PATH OR 4D MATRIX, OPTIONAL,
 %               DEFAULT = generated below from ASL4D)
-%   PWI3D   - likewise, see explanation below. If this is provided, it will override the PWI4D calculation below (STRING OR 4D MATRIX, OPTIONAL,
+%   PWI3D      - likewise, see explanation below. If this is provided, it will override the PWI4D calculation below (PATH OR 4D MATRIX, OPTIONAL,
```
These shouldn't be here. The JSON management is in xASL_io_ASLSubtractionAveraging, not in xASL_im_ASLSubtractionAveraging.

jan-petr replied:
see answer below

jan-petr replied:
actually, it doesn't hurt in any way to have it here - it's a plus, rather than a minus to me :)

jan-petr replied:
I removed this. Now use the IO version for this purpose... a659dc33

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m:
```diff
@@ -147,6 +151,16 @@
 
 % We skip the computation of PWI4D if it is provided
 if nargin>=3 && ~isempty(PWI4D)
+	if ischar(PWI4D) && xASL_exist(PWI4D, 'file')
```
These shouldn't be here. The JSON management is in xASL_io_ASLSubtractionAveraging, not in xASL_im_ASLSubtractionAveraging.

Same for other instances below

jan-petr replied:
Actually - not really. xASL_io_ASLSubtractionAveraging loads and saves everything...
We only want to load and average stuff, not to save everything. 

So thats why xASL_im_ASL was better. Actually, xASL_im_ASL is used already independently on many places with ASL4D as input. And here I only needed to average things so that I could run it for despiked_ASL. I only wanted a correct average of despiked.nii and not create and save despiked_PWI3D etc.

HenkMutsaerts replied:
The way I designed it, you have the following choices (on purpose):

1. We always use xASL_io_ASLSubtractionAveraging, because you always load and save NIfTIs/JSONs. We don't want to put stuff back to memory anymore.
2. For (despiked_)ASL4D only, we currently have an exception, because we want to be able to create a temporary PWI from the raw ASL4D with motion application, as temporary PWI.

What you changed, is that you can also input non-ASL4D paths to xASL_im_ASLSubtractionAveraging, which is a behavior we don't want. So please revert this, and use one of the 2 options above.

Do you agree with this logic?

jan-petr replied:
Yes. As above - removed in a659dc33

jan-petr replied:
Note that I have changed this behavior on all occurrences, so I closed all the other conversations because I change it everywhere how you asked.

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_MergePWI4D.m:
```diff
@@ -0,0 +1,101 @@
+%   x                   - structure containing fields with all information required to run this submodule (REQUIRED)
+%
+% OUTPUT: 
+%   path_PWI4D          - Path to the new merged file
```
file -> NIfTI (same throughout)

jan-petr replied:
Agree. Will change everywhere...

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_MergePWI4D.m:
```diff
@@ -0,0 +1,101 @@
+	error('Empty input parameters');
+end
+
+fprintf('Concatenating PWI4D for sessions:');
```
I would print the name of the sessions/runs before issuing the errors below.

jan-petr replied:
They will be printed below if there are no errors.
And errors anyway point out sessions without needing to know the list.

I can of course add it, just asking if it is worth it.

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_MergePWI4D.m:
```diff
@@ -0,0 +1,101 @@
+
+		% Error with a missing JSON
+		if isempty(jsonPWI4Dcurrent)
+			error(['Cannot concatenate sessions for subject ' x.SUBJECT ': session '  x.modules.asl.sessionsToMerge{iSession} ' has missing JSON information.']);
```
This is probably a good idea though users of older ExploreASL versions will get this error without knowing what to do.

I would perhaps still merge the NIfTIs (why not)? And issue a warning.
An error is also fine. This reduces the flexibility but increases safety.



"has missing JSON information" -> is too general, information about a JSON can be anything.
I would say "does not have a JSON sidecar"

jan-petr replied:
if we merge and don't have JSONS, then we are fucked later on :)

I'm fine doing that, but we should maybe release version 2.0 and keep only a reasonable compatibility... If you process part of ASL module with version 1.11.0 and second part of ASL module with 1.12.0, it will crash. But I thought that we agreed upon previoulsy that nobody should do a dirty thing like this :)

HenkMutsaerts replied:
Ok, then only:

"has missing JSON information" -> is too general, information about a JSON can be anything.
I would say "does not have a JSON sidecar"

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_MergePWI4D.m:
```diff
@@ -0,0 +1,101 @@
+			% For the first session, we take both JSON and NII as they are
+			imPWI4DConcatenated = imPWI4Dcurrent; 
+
+			% Note the we don't convert to Legacy and work with BIDS format directly
```
the -> that

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_MergePWI4D.m:
```diff
@@ -0,0 +1,101 @@
+
+		% Print the name of the session being added
+		if iSpace == 1
+			fprintf([' ' x.modules.asl.sessionsToMerge{iSession}]);
```
So I would move the fprintf of the current session above, so it always prints this before trying anything, easier to read any error.

jan-petr replied:
as said above - it will still work. but i can move that without any problem if you think it's worth it. 

Just checking... not oppossing.

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_MergePWI4D.m:
```diff
@@ -0,0 +1,101 @@
+				jsonPWI4DConcatenated.PostLabelingDelay = [jsonPWI4DConcatenated.PostLabelingDelay; jsonPWI4Dcurrent.PostLabelingDelay];
+				jsonPWI4DConcatenated.LabelingDuration = [jsonPWI4DConcatenated.LabelingDuration; jsonPWI4Dcurrent.LabelingDuration];
+			else
+				error(['Cannot concatenate sessions for subject ' x.SUBJECT ' as session '  x.modules.asl.sessionsToMerge{iSession} ' has a different matrix size than the previous sessions.']);
```
Note that a single subject can have multiple session combinations to merge, e.g. ASL_1 ASL_2 and ASL_3 ASL_4. So you cannot say that you "cannot concatenate sessions for subject ..."

Or are we implementing only a single merging list option?

In the error, I would say "cannot concatenate sessions X and X (subject YY)" which makes it specific, not general about the subject as a whole.

jan-petr replied:
Agree. Will print the entire list and point out the wrong session.

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_MergePWI4D.m:
```diff
@@ -0,0 +1,101 @@
+		end
+
+		% Print the name of the session being added
+		if iSpace == 1
```
!!! iSpace == 1  -> iSession == 1

jan-petr replied:
That is correct. 
For space 1, session 1 it prints ASL_1, then for space 1, session 2 it prints ASL 2.

For space 2 it doesn't do anything.

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_MergePWI4D.m:
```diff
@@ -0,0 +1,101 @@
+		end
+	end
+	% Create the specific new name
+	[fPath, fName, fExtension] = xASL_fileparts(pathPWI4D{iSpace});
```
I would hardcode fExtension here, as otherwise this can vary depending on taking the fileparts from session 1 or 2 or n in the list, and same for the 2 spaces.

I would hardcode as .nii, because the merging is something we do within the processing, e.g. we will reuse the .nii before zipping it at the end of the pipeline.

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_MergePWI4D.m:
```diff
@@ -0,0 +1,101 @@
+                imPWI4DConcatenated = cat(4, imPWI4DConcatenated, imPWI4Dcurrent);
+
+				% Concatenate the JSON info
+				jsonPWI4DConcatenated.EchoTime = [jsonPWI4DConcatenated.EchoTime; jsonPWI4Dcurrent.EchoTime];
```
I would harmonize this with what we have in the rest of our PWI handling:

1. check if the JSON exists (you already have this)
2. check if PLD & LD exists (and merge them, or issue warning/error)
3. check if TE exists (and merge them, or issue warning/error)

Note that you might merge ASLs without knowing the TE which can be perfectly fine (this is not only about DEBBIE). And you don't want to handle PLDs without knowing the LD or vice versa, because they are interdependent.

HenkMutsaerts commented on Functions/Quantification/xASL_quant_ASL.m:
```diff
@@ -57,20 +59,24 @@
     error('Illegal x structure');
 end
 
+if nargin<6 || isempty(bSaveCBF4D)
```
narging 6 5 6, are we missing 4 ?

jan-petr replied:
Sorry, I only modified your code. Will check

HenkMutsaerts commented on Functions/Quantification/xASL_quant_ASL.m:
```diff
@@ -45,10 +46,11 @@
-% Copyright (c) 2015-2023 ExploreASL
+% Copyright (c) 2015-2024 ExploreASL
 
+%% #1543 In case we save as CBF4D or bUseBasil then we don't average. Otherwise, for non-basil non-cbf4D, we have to average PWI4D before quantification - perhaps in this subfunction.
```
ToDo

jan-petr replied:
I know... that's still in the list of issues... No need to report.

HenkMutsaerts commented on Functions/Quantification/xASL_quant_ASL.m:
```diff
@@ -86,6 +92,14 @@
 
+% Load ASL single PWI
+[PWI4D, jsonASL] = xASL_io_Nifti2Im(PWI4D_Path); % Load CBF nifti
+jsonASL = xASL_bids_parms2BIDS([], jsonASL, 0); % BIDS to Legacy conversion
```
This should be inside xASL_io_ReadNifti (called by Nifti2Im)

HenkMutsaerts commented on Functions/Quantification/xASL_quant_ASL.m:
```diff
@@ -163,10 +177,8 @@
-        % #1543 We may want to move the NaN extrapolation here
-
-		[PWI, ATT, ABV, Tex] = xASL_quant_FSL(x.P.Path_PWI4D, x); % #1543 Quick and dirty fix for the NaN extrapolation at the start of xASL_quant_FSL
+		% We pass the path to the image and do all the Image and JSON reading inside the function
```
It seems less "clean" because we deal with a lot of other parameters for which we load the JSON, x, etc outside of xASL_quant_FSL (things that are equal for BASIL & xASL-quantification). So you load JSON and x parameters for the non-specific quantification (background suppression, vendor-specific, etc) separately from loading the JSON and x parameters for the specific quantification (what BASIL does differently from xASL). 

So it might be easier in the future to change this, also when we revamp xASL_wrp_Quantify?

jan-petr replied:
I don't understand. We provide x-struct with general params. And path (NII+JSON) for all specific ones. That seems to be the best flexible solution.

HenkMutsaerts replied:
What I'm saying, is that in the future, x contains only the settings, and all data-specific parameters (which background suppression, which vendor etc) are only in the JSON.
Currently, we are almost there. We still have x.Q.Background* , x.Q.vendor* , just a few ones.

jan-petr replied:
Yes - I fully agree (x will have settings. JSON will have specifics).

So that's why I pass PWI4D_Path (which has the JSON as well) and x-struct. 
So the code should be correct and not require any changes?!

HenkMutsaerts replied:
Not completely, is what I mean. x.Q.vendor and x.Q.Background... are examples of settings that belong in the JSON but that are in x. So we don't have this completely.

So now, or in the future revamp that I pointed to above, we remove x.Q completely. Agree?

jan-petr replied:
Sorry for confusion. Yes, I agree, x.Q.vendor should be removed in the future. 

I meant that apart from this (x.Q.vendor etc) which will be done later, we have everything correct :)

You can resolve this after reading.

HenkMutsaerts commented on Functions/Quantification/xASL_quant_FEAST.m:
```diff
@@ -65,9 +61,12 @@ function xASL_quant_FEAST(x)
+		PLD{iSession} = x.Q.uniqueInitial_PLD + SliceReadoutTime(SliceNumber);
+		CBF{iSession} = CBF{iSession}./(exp(PLD{iSession}./x.Q.BloodT1) / (2.*x.Q.LabelingEfficiency.*x.Q.BloodT1 .* (1- exp(-x.Q.uniqueLabelingDuration./x.Q.BloodT1)) ));
+	else
+		error('FEAST can only handle sequences with single PLD and labeling duration.');
```
a single PLD

HenkMutsaerts replied:
I would have done the unique check separately a few lines up, which would have kept the PLD CBF lines equal (so smaller diff). Just saying :)

jan-petr replied:
I've spotted the error here. that's why. But agree to move this.

HenkMutsaerts commented on Functions/Quantification/xASL_quant_M0.m:
```diff
@@ -149,7 +149,13 @@
+			if isfield(x.modules.asl, 'bTimeEncoded') && x.modules.asl.bTimeEncoded == 1
+				% For Time-Encoded, we can't take the longest PLD, but the longest PLD in PWI4D. 
+				% This is by definition the last PLD in the vector, as typically the first corresponds to the control image and can be ignored
+				NetTR = x.Q.LabelingDuration(end) + x.Q.Initial_PLD(end) + SliceReadoutTime(SliceIM);
```
This needs to be replaced by the read PLD & LD from a PWI4D, because we don't know if PWI4D has different largest LD/PLD than the original sequence (e.g., if parts have been removed manually or in despiking).

But we can make this a future issue.

jan-petr replied:
Good point. In case there's PWI4D, we'll read parameters from there.

No point to wait - fixed here: d3271b25

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_ProcessM0.m:
```diff
@@ -98,6 +98,7 @@ function xASL_wrp_ProcessM0(x)
 % Copy existing M0 for processing, in single precision, we can skip the motion correction here
 % Load and scale the M0 for the ASL-M0 volume changes
 [imM0, jsonM0] = xASL_io_Nifti2Im(x.P.Path_M0);
+jsonM0 = xASL_bids_parms2BIDS([], jsonM0, 0); % BIDS to Legacy conversion
```
See previous

jan-petr replied:
Actually - with all JSON reading everywhere - we are applying this explicitly. So adding it to Nifti2IM will IMHO really mess things up.

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_ProcessM0.m:
```diff
@@ -132,7 +133,7 @@ function xASL_wrp_ProcessM0(x)
 
 % Save the resampled mean M0
-xASL_io_SaveNifti(x.P.Path_M0, x.P.Path_rM0, imM0, 32, 0);
+xASL_io_SaveNifti(x.P.Path_M0, x.P.Path_rM0, imM0, 32, 0, [], 0, xASL_bids_parms2BIDS(jsonM0, [], 1));
```
Would also incorporate xASL_bids_parms2BIDS in xASL_io_SaveNifti. You have made this into a nicely single function that manages legacy to BIDS conversion and back. Best to add a boolean and default to what we decide, rather than manually insert this function in saving/loading.

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_Quantify.m:
```diff
@@ -73,9 +67,19 @@ function xASL_wrp_Quantify(x, PWI4D_Path, pathOutputCBF, M0Path, SliceGradientPa
+	bSaveCBF4D = false;
+end
+
+if x.modules.asl.bUseBasilQuantification && bSaveCBF4D
```
Why not? That should be easy right

jan-petr replied:
BASIL doesn't produce CBF4D. It does fitting to all data.
To have CBF4D, you would have to run BASIL separately for each dynamics. 

So it would be easy, but it would take enormous amount of time. Same why we don't do BASIL in standard space.

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_Quantify.m:
```diff
@@ -98,7 +102,9 @@ function xASL_wrp_Quantify(x, PWI4D_Path, pathOutputCBF, M0Path, SliceGradientPa
 xASL_delete(pathOutputABV);
 
-if x.Q.bUseBasilQuantification
+% For BASIL, only native data are processed and standard space data are not directly quantified, but only transformed
```
Please stop using the word "data" if it can be replaced by something more specific :) It sounds like you're a machine learning guy that doesn't care whether your data are images or just scalar values, other than perhaps the dimensionality :)

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -401,47 +401,53 @@
-%% This is not part of the quantification, just as HadamardDecoding is also not part of the quantification, it is a resorting/reconcatenation of image & parameter matrices
-
+%% % Use either original or motion estimated ASL4D
+% #1543 Make sure that despiked is used instead of normal ASL in case it exists - to be solved here rather than inside Quantify
```
Good point. I would remove the #1543 here.

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -401,47 +401,53 @@
 iState = 8;
 if ~x.mutex.HasState(StateName{iState}) && x.mutex.HasState(StateName{iState-4})
 
+	%% Run merging if needed as first function before any quantification, as quantification itself should disregard the data source
```
data -> image

HenkMutsaerts replied:
needed -> requested

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -401,47 +401,53 @@
 if ~x.mutex.HasState(StateName{iState}) && x.mutex.HasState(StateName{iState-4})
 
+	%% Run merging if needed as first function before any quantification, as quantification itself should disregard the data source
+	if x.modules.asl.bMergingSessions == 1
```
You can remove == 1

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -401,47 +401,53 @@
 
+	%% Run merging if needed as first function before any quantification, as quantification itself should disregard the data source
+	if x.modules.asl.bMergingSessions == 1
+		[path_PWI4D, path_PWI4D_Pop] = xASL_im_MergePWI4D(x);
```
I wonder if we should not change x.P.Path_PWI4D rather than creating a new variable.
In the QC part, don't we want to get QC from the same NIfTIs as we did the quantification from?

I can't remember what we decided. But we 

EITHER 
1. Only quantify & do QC on the merged PWI4D
OR
2. Both quantify & do QC on the non-merged & merged PWI4D

Now we only quantify the merged PWI4D but only do QC on the non-merged PWI4D, that's not consistent and confusing to the user.

So if we want (1), I would change x.P.Path_PWI4D rather than create a new variable.
This will propagate to the QC wrapper which will perform visual and quantitative QC on the merged PWI4D.

jan-petr replied:
changing x.P.Path_PWI4D is a bad idea as this is a standardized parameter. We can't redirect it.

HenkMutsaerts replied:
We do that on several instances, eg with despiked_ASL4D. I understand it's less appropriate, but we need a parameter to propagate to do the QC.

jan-petr replied:
With despiked_ASL4D, we create PWI4D based on despikedASL4D instead of based on ASL4D.
So I'm fine creating a new parameter called Path_PWI4D_REALLY_USED_WHATEVER. But not fine changing an automatically generated path parameter that should always point to PWI4D.nii.
so this path_PWI4D stays, but can be saved under a special name into x-struct.

HenkMutsaerts replied:
Yes, I agree with your solution. We should in fact do this with despikedASL4D as well, instead of changing the fixed x.P.Path_...

We could call it x.P.Path_PWI4D_used
(and similarly x.P.Path_ASL4D_used; but let's keep despikedASL4D for now perhaps;)

alternative = x.P.Path_PWI4D_custom

and we know that x.P.*_used can be overwritten, whereas x.P.* always stays default. Good idea?

jan-petr replied:
So right now we have `path_PWI4D`, shall I change that to `path_PWI4D_used`? I'm fine with this.

Just not sure if it's worth it to save it to x.P as we anyway pass this as a function argument to wrp_Quantify and we have to pass it as we sometimes use wrp_Quantify in native and sometimes in standard space. So I'm just not sure if saving to x.P would bring any benefit.

HenkMutsaerts replied:
Yes let's do path_PWI4D_used, just to clarify that this has a different meaning. I thought about x.P. because we will need this in multiple functions, not only quantification but also visualization, QC, etc.

jan-petr replied:
I checked all occurences of x.P.Path_PWI4D and x.P.Pop_Path_PWI4D. And the only relevant place using PWI4D that could be altered here is xASL_wrp_VisualQC_ASL.

So I save stuff to x.P.Path_PWI4D_used and then get it from there for the QC. All other occurences are handled by passing the correct path as function parameter.

Done in 78ed26a4

HenkMutsaerts replied:
OK great; just note that there may be extra occurrences coming as I will implement more QC steps using this PWI4D (probably in xASL_wrp_VisualQC_ASL). So I'm fine for now.

Please change

if x.modules.asl.bMergingSessions == 1

TO

if x.modules.asl.bMergingSessions

And I don't get why we should have both

path_PWI4D
path_PWI4D_Pop

and 

x.P.Path_PWI4D_used & x.P.Pop_Path_PWI4D_used (note that we need both in visual QC! even it is not there yet)

So I would just use 

x.P.Path_PWI4D_used & x.P.Pop_Path_PWI4D_used

here as well. Adding a description that we also use this in xASL_wrp_Quantify and xASL_wrp_VisualQC.

The less variables the better.

jan-petr replied:
1. "if x.modules.asl.bMergingSessions" agree, but that has already been fixed :)
2. "path_PWI4D" - agree, but you looked at a "intermediate" commit. We already use only x.P.Path_PWI4D_used in xASL_module_ASL.
3. Yes - of course, more occurrences are coming - that's fine.
4. Agree - added a comment that the "used" variable is important for xASL_wrp_VisualQC.

4 done in 696ad616

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_Quantify.m:
```diff
@@ -109,59 +115,19 @@ function xASL_wrp_Quantify(x, PWI4D_Path, pathOutputCBF, M0Path, SliceGradientPa
-    ASL_parms = xASL_adm_LoadParms(x.P.Path_ASL4D_parms_mat, x);
+% Load ASL single PWI
+[~, jsonASL] = xASL_io_Nifti2Im(PWI4D_Path); % Load CBF nifti
+jsonASL = xASL_bids_parms2BIDS([], jsonASL, 0); % BIDS to Legacy conversion
```
to be done inside xASL_io_NIfti2im->ReadNifti

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_Quantify.m:
```diff
@@ -109,59 +115,19 @@ function xASL_wrp_Quantify(x, PWI4D_Path, pathOutputCBF, M0Path, SliceGradientPa
-    x.Q.EchoTime_PWI4D = JSON.Q.EchoTime;
-    x.Q.InitialPLD_PWI4D = JSON.Q.Initial_PLD;
-    x.Q.LabelingDuration_PWI4D = JSON.Q.LabelingDuration;
+if isempty(jsonASL)
```
This is a fix for backward compatibility only.
How about we:

1. issue a warning about the missing JSON
2. try to find a legacy x.P.Path_ASL4D_parms_mat; if we find it we use this and x.Q.*, still issuing another warning

Otherwise, we have to say here how a user can run this part without having a JSON, for backward compatibility. Or simply state that we break backward compatibility ;)

jan-petr replied:
we can't find this in x.Q or in parms_mat - cause that wouldn't be valid for despiked data. 
We rewrote this function to be general and accept any input, so here, we can't anymore use x.Q.

HenkMutsaerts replied:
Yes, it's a backward compatibility issue only. I agree that we should break the backward compatibility. Shall we at the start of xASL_module_ASL do a check for data validity (e.g., issue a warning for missing JSON that this needs fixing first, and detecting _parms.mat and saying that this is old and needs conversion)?

If we do this "backward compatibility/data validity" check at the start of xASL_module_ASL, we can remove these checks from the rest of the wrappers.

jan-petr replied:
For missing JSONs and non-missing _parms.mat for ASL4D.json at the start of the ASL module, we should indeed make a new check and make a new issue for this.

However, this check here, is not about having old data. This is just a random check that the PWI4D.json was correctly calculated and nobody removed it :D So we can make this new issue, but also keep this check here to help users deal with random errors. Though we can also remove it as not really necessary.

HenkMutsaerts replied:
I see, then we need to change the comment:

% If jsonASL is missing, this could mean either:
% A. this is old derivatives data, where we didn't use json sidecars yet
% B. Somebody changed the PWI4D.nii but accidentally removed the json sidecar
% In both cases, we have to throw an error. Because what we have in x.Q cannot be used, this is a leftover from the start of the pipeline (from the rawdata unsubtracted volumes).

And remove the part about "we could have recalculated ... in the previous functions." -> because that is confusing to the user imho

jan-petr replied:
Fully agree. Changed that inside: d3271b25

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_Quantify.m:
```diff
@@ -244,12 +194,14 @@ function xASL_wrp_Quantify(x, PWI4D_Path, pathOutputCBF, M0Path, SliceGradientPa
 if strcmpi(x.Q.M0,'separate_scan')
-    M0_parms = xASL_adm_LoadParms(x.P.Path_M0_parms_mat, x);
+	[~, jsonM0] = xASL_io_Nifti2Im(x.P.Path_M0);
+	jsonM0 = xASL_bids_parms2BIDS([], jsonM0, 0);
```
See above

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_Quantify.m:
```diff
@@ -482,7 +434,7 @@ function xASL_wrp_Quantify(x, PWI4D_Path, pathOutputCBF, M0Path, SliceGradientPa
 %% ------------------------------------------------------------------------------------------------
 %% 11.  Create standard space masked image to visualize masking effect
-if xASL_exist(x.P.Pop_Path_qCBF, 'file') && (strcmp(pathOutputCBF, x.P.Pop_Path_qCBF) || x.Q.bUseBasilQuantification)
+if xASL_exist(x.P.Pop_Path_qCBF, 'file') && (strcmp(pathOutputCBF, x.P.Pop_Path_qCBF) || x.modules.asl.bUseBasilQuantification) && ~bSaveCBF4D
```
Weird construction?

This should not be BASIL-dependent, BASIL gives the same qCBF image and masking should still be done.

Also: bSaveCBF4D means that you save both CBF and CBF4D, so this doesn't preclude creating the standard space masked image?

jan-petr replied:
bSaveCBF4D -> see above.

HenkMutsaerts replied:
bSaveCBF4D -> I understand that BASIL doesn't output this. But then we should issue a warning about this, and say that for saving CBF4D you need the non-BASIL quantification

masking: here we created a masked version of the standard space qCBF, which we can still (and should still) do for BASIL/FABBER output

jan-petr replied:
1. We already have such a warning in xASL_module.ASL line 434:
    warning('Saving CBF4D was requested but not implemented yet for multi-PLD or multi-TE, skipping');
    So we don't need a second warning here, just a safety check.
    
 2. OK, mask for BASIL output -> please make a more specific comment of what to do and where.

HenkMutsaerts replied:
1:
My point is: we have several checks for setting combinations. If there is an invalid setting combination, we issue a warning, and change the setting. We do this ideally for all settings at a single location, but at least we don't do it twice. So choose if you want this here or in xASL_module_ASL, and change throughout.
The reason is that it is nicely modular/readable. If we do a setting check at a location where we only throw a warning but not change the setting, and not give an explanation, this is confusing and makes our code less readable. Let's have this all at a single location.

So we have the A setting check B warning C setting change at a single location, and remove this elsewhere.

I would probably be in favor of having this for all combinations at the quantification start itself, because it is closest to the action. I know that for quantification parameters like PLD, TE, LD we have this at the start of xASL_module_ASL, but that is because we need those settings throughout the whole pipeline. But I would also be fine to move these settings to the start of xASL_module_ASL, but at least at a single location.


2: Here we create a qCBF_masked. If BASIL is run, we still need this (but then create this when qCBF is already created, this should be similar to without BASIL). When we have bSaveCBF4D, we ALSO save qCBF so also qCBF_masked. So saving qCBF4D doesn't exclude saving qCBF and qCBF_masked.

-> qCBF_masked is used for population module-> CBF ROI values, as it removes vascular, whereas qCBF is used for spatial CoV ROI values. It is nice to keep these NIfTIs so users can see what was masked out for the ROI values (but not necessary for archiving)

jan-petr replied:
1. OK. I get your point now. And agree that a single location for warnings is a must. Actually, the best location is probably the start of xASL_module_ASL in the function where we check all parameters. I'll make the check there, input the warning and then switch this setting of. In the original place of the code, I'll leave only safety catch without an explanation for the user - just in case somebody messes up. Done in 44aa553b

2. Irrespective of bSaveCBF4D, we always save qCBF and qCBF_masked - this happens in the step above. Without BASIL we do it for both native and standard. With BASIL, we do it for native only, but then transform this to standard. At least it should be like this unless there is some bug. So things should actually already work as you want. If you feel that some intermediate image is delete, then I agree that we don't have to delete it. Not sure if this answers your comments, but I don't really understand what the issue is here. 

This is done, so you can check and resolve.

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_RegisterASL.m:
```diff
@@ -314,13 +314,19 @@
+jsonFields.Initial_PLD = xSubtracted.Q.InitialPLD_PWI;
+jsonFields.LabelingDuration = xSubtracted.Q.LabelingDuration_PWI;
+jsonFields.EchoTime = xSubtracted.Q.EchoTime_PWI;
+xASL_io_SaveNifti(x.P.Path_despiked_ASL4D, x.P.Path_mean_PWI_Clipped, PWI, 16, false, [], 0, xASL_bids_parms2BIDS(jsonFields, [], 1));
```
Isn't this json saving done in xASL_io_ASLSubtractionAveraging?

jan-petr replied:
in IO yes, but in IM no...

HenkMutsaerts replied:
yes, so we don't use xASL_im_ASLSubtractionAveraging here but we use xASL_io_ASLSubtractionAveraging instead right?

jan-petr replied:
You were right. Using the IO function a659dc33

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_RegisterASL.m:
```diff
@@ -314,13 +314,19 @@
+	jsonFields.Initial_PLD = xSubtracted.Q.InitialPLD_Control;
+	jsonFields.LabelingDuration = xSubtracted.Q.LabelingDuration_Control;
+	jsonFields.EchoTime = xSubtracted.Q.EchoTime_Control;
+    xASL_io_SaveNifti(x.P.Path_despiked_ASL4D, x.P.Path_mean_control, Control, 16, false, [], 0, xASL_bids_parms2BIDS(jsonFields, [], 1));
```
Same here

jan-petr replied:
same as above - that wouldn't work with this extra saving. and IO is not applicable here as explained on another place

HenkMutsaerts replied:
Yes, IO is applicable here, I wrote that here right? I don't understand why you removed it.
If we don't want to save, you can use IM. But if we want to save, we should use IO. That is the whole purpose of IO, to have this harmonized rather than writing these same 5 lines every instance

jan-petr replied:
You were right. Using the IO function. a659dc33

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_ResampleASL.m:
```diff
@@ -245,24 +245,29 @@ function xASL_wrp_ResampleASL(x)
-    [PWI4D, json] = xASL_io_Nifti2Im(x.P.Pop_Path_PWI4D);
+    [PWI4D, json_PWI4D] = xASL_io_Nifti2Im(x.P.Pop_Path_PWI4D);
+	% Convert JSON from BIDS to Legacy
+	json_PWI4D = xASL_bids_parms2BIDS([], json_PWI4D, 0);
```
See above

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_ResampleASL.m:
```diff
@@ -245,24 +245,29 @@ function xASL_wrp_ResampleASL(x)
-    PWI4D_statsIndices = json.Q.Initial_PLD==max(x.Q.uniqueInitial_PLD);
-    %% #1543 For some reason this didn't work for the combination shortest TE & longest PLD
-    %% Better to repeat this for every TE-PLD-LD combination 
+    %% Use the positions for the earliest echo, and the latest PLD-labdur
```
I thought we did this with the TE-PLD-LD weighting function that I wrote?

I am confused why we don't also use this here

jan-petr replied:
We did weighted for registration. This is for single-PLD quantification of multi-PLD data. For that we can't weight

HenkMutsaerts replied:
Are you sure we can't weight? But more importantly: always choosing the last PLD will not necessarily be the best option for single-PLD quantification of multi-PLD data, as you pointed out before. Shall we make a new issue out of this (with low priority), then I agree with you here.

jan-petr replied:
For quantification, we need to know the exact parameters. Doing a single-PLD quantification on averaged multi-PLD data is a bad idea, so we really have to choose a single-PLD or do a proper multi-PLD quantification rather than weighted averaging.

Choosing the last PLD is indeed a bad idea, that's why I rewrote the code to choose the PLD closest to 2000 (which was our top choice for the weighting as well). 

I'm fine choosing a slightly different strategy than 2000ms. Any suggestions? Cause very likely, the solution will be very simple and we can make the change immediately without making a new issue.

HenkMutsaerts replied:
Ah OK I see the thinking difference here. The PWI created here is just a visual reference, will and should never be used for quantification. Only the PWI or CBF created in xASL_wrp_Quantify should be used for that. So if we perform single-PLD quantification of multi-PLD data, this will happen in xASL_wrp_Quantify.

xASL_wrp_Resample is for:
1. PWI4D that will later be used for quantification (after potential merging/manual adaptations)
2. Other reference images for PVC, visual QC, etc. For this, we want all available SNR we can get, similar to the temporary PWI created in xASL_wrp_RegisterASL. This is why we want to do the same weighting here as we did in xASL_wrp_RegisterASL (probably by a single weighting function).

Historically, we created SD, SNR, and other QC stuff in xASL_wrp_Resample. The question is where we keep this, or if we move this to after quantification. Because we probably want this to be AFTER the merging/manual removal right? (please also think about Mervin merging two single-PLD, one at the start and one at the end of the scan protocol, not only think about a single exotic DEBBIE sequence, but think generalized ;)

jan-petr replied:
I'm of course thinking of other users and other sequences. The PWI we are choosing here is used to calculate SD and SNR, not for visualization. If you use weighting, then SD and SNR will have no value.

I guess you can resolve this.

HenkMutsaerts reviewed (changes requested):
Supernice! few points

jan-petr committed:
#1543 JSON conversion with NIfTI saving/loading

jan-petr committed:
#1543 Fixing style and minor things

HenkMutsaerts commented on External/SPMmodified/xASL/xASL_io_Nifti2Im.m:
```diff
@@ -64,7 +70,12 @@
         else
             % Load NIfTI and JSON sidecar
-            [nii, ~, json] = xASL_io_ReadNifti(niftiIn);
+			if nargout < 2
```
This looks complicated. I would remove the nargout. And if you would even need it, you would just disable bBIDS2Legacy instead of repeating the command. We default JSON to empty. This is how I programmed it.
So it always gets outputted whatever you do, whereever it crashes. And you can easily catch an empty JSON variable outside this function. You cannot catch a non-existing parameter.

jan-petr replied:
It might look complicated, but it's the correct thing to do and makes no sense to do it otherwise. 
If the user doesn't request json, then it makes no sense to waste resources in reading it.

This is not about catching errors or initialization, it is about skipping the calculation of certain things that the user didn't request. This makes the code more faster and less time complex, which is especially helpful for image reading that takes unnecessary resources. Or 90% of cases of NII reading doesn't require the JSON. It's just an optimization of code without any adverse effects of caveats

HenkMutsaerts replied:
Yes also here I agree, as this is a basic function that is called a lot

HenkMutsaerts commented on External/SPMmodified/xASL/xASL_io_ReadNifti.m:
```diff
@@ -88,10 +95,14 @@
-if exist(pathJson)
-    json = xASL_io_ReadJson(pathJson);
+%% Load JSON sidecar if it exists and if the output is required
+if nargout > 2
```
Also here. Let's remove the use of nargout, it is superfluous.
I understand this way possibly speeds up stuff but you make the code more conditional so less flexible/transparent.

I don't feel strong about this one though so feel free to resolve this if you want to keep it as is

jan-petr replied:
Yes - it doesn't need to be there, I agree. But it's to speed up things, limit disc access. Makes no huge difference in the end of course, but it's a good practice to avoid doing unnecessary things.

So given that we have clear comments in the code about this, I'd say that it doesn't make the code more complicated. 

So I'd indeed keep this as I feel strong about this. We have improved our code a lot to be readable, but we have a lot to work on on optimal processing speed etc. This is the first small step :) You can resolve this - I wanted you to hear this explanation...

HenkMutsaerts replied:
OK fine. I understand this for a basic function that can be called 1000's times, you make a good point here as the speed increment can be is relatively major. I just would not like to sacrifice code lines (every extra if-statement adds to code complexity) for slight speed increments, e.g., I would not like `if nargout` for xASL_wrp_ or other functions where the speed increment is relatively minor.

You can resolve this, unless you disagree with me but then let's discuss later/in whatsapp to find a general consensus. I agree with you for these kind of basic functions!

jan-petr replied:
Of course I agree.

We aim for a good balance between speed and complexity of course.

So tiny speed-up tricks are only used in function called a lot where tiny improvements and high optimization matters. In bigger functions with large overhead, we don't bother with such things of course.

So I fully agree with you, it would not even occur to me to do it differently :)

You can resolve after reading.

HenkMutsaerts commented on External/SPMmodified/xASL/xASL_io_SaveNifti.m:
```diff
@@ -298,15 +305,20 @@ function xASL_io_SaveNifti(pathOrigNifti, pathNewNifti, imNew, nBits, bGZip, cha
-    end
+	end
+
+	% b. Second, we convert provided JSON fields to BIDS if necessary
```
"if necessary" -> "if requested"

HenkMutsaerts commented on External/SPMmodified/xASL/xASL_io_SaveNifti.m:
```diff
@@ -298,15 +305,20 @@ function xASL_io_SaveNifti(pathOrigNifti, pathNewNifti, imNew, nBits, bGZip, cha
+	end
+
+	% b. Second, we convert provided JSON fields to BIDS if necessary
+	if bLegacy2BIDS && ~isempty(JsonFields) && length(fieldnames(JsonFields))
```
length(fieldnames(JsonFields)) -> ~isempty(fieldnames(JsonFields))

HenkMutsaerts replied:
length(fieldnames(JsonFields)) -> ~isempty(fieldnames(JsonFields))

jan-petr committed:
#1543 Minor fixes based on comments

jan-petr committed:
#1543 Proper use of xASL_io_ASLSubtractionAverage

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_RegisterASL.m:
```diff
@@ -314,20 +314,9 @@
-	jsonFields.EchoTime = xSubtracted.Q.EchoTime_Control;
-    xASL_io_SaveNifti(x.P.Path_despiked_ASL4D, x.P.Path_mean_control, Control, 16, false, [], 0, jsonFields, true);
-end
+saveWhichNifti{1,1} = 1; saveWhichNifti{1,2} = x.P.Path_mean_PWI_Clipped;
```
saveWhichNifti needs to be a list right? Now you overwrite it

HenkMutsaerts commented on Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m:
```diff
@@ -174,15 +165,6 @@
-		x.Q.LabelingDuration_Control4D = json.Q.LabelingDuration;
-	end
-
     if ~isnumeric(Control4D)
```
I would sort the nargins numerically

jan-petr replied:
Sure - that was you who wrote that...

Fixed. You can resolve

HenkMutsaerts commented on Functions/Quantification/xASL_quant_ASL.m:
```diff
@@ -105,10 +105,17 @@
     fprintf('%s\n','We skip the scaling of a.u. to label intensity');
 	% In case we don't run quantification, we still need to average the PWI image
-	PWI = xASL_im_ASLSubtractionAveraging(x, [], PWI4D_Path);
+	xASL_io_ASLSubtractionAveraging(x, {2, PWI3D_Path}, 0, [], PWI4D_Path);
```
Why don't you replace this with 

PWI = xASL_im_ASLSubtractionAveraging(x, [], xASL_io_Nifti2Im(PWI4D_Path));

I say this, because in the future I would like to simplify this:
1. load PWI4D
2. run BASIL or not
3. do the scalings
4. average the 4th dim (which would have no effect if it was already 1)
-> do you agree that we do this in #1572 ?

jan-petr replied:
Henk. You have just asked me to replace the IM by IO. Now you are asking me to switch back.

Moreover, we can't do this, because by loading PWI4D, we don't have information anymore about the PLDs as these are in the JSON. We can't use x.Q, because PWI4D here can be normal, despiked or merged.

HenkMutsaerts replied:
I have asked you to replace IM by IO in general, where it makes sense.

AFAIK at this point, we have already loaded the NIfTI and JSON earlier in the function, right?

See my point to reorder/simplify xASL_quant_ASL. If functionally everything works, we can keep it as it is, and do this in the later revamp.

jan-petr replied:
Regarding xASL_io_ASLSubtraction - we have to use IO and not IM here, because we need to update PWI3D based on a new PWI4D. So this is not about loading JSON, but about saving new JSON+NII, so the IO has to be used as IM wouldn't save anything.

So we can't modify the code here. And we can do the rest in #1572.

HenkMutsaerts replied:
There is no new PWI4D. This is xASL_quant_ASL, which at the start of the function reads NIfTI and JSON, and not here.

There is a wrapper outside this function that does all the management. But this function should not do this, we are not saving intermediate steps here.

We can do this later. But again:

we have whatever wrappers that:
1. load PWI4D
2. load JSON
3. Manage basil/xASL quantification options bla bla
4. start xASL_quant_ASL -> which does the quantification
5. Save CBF

We do not save PWI3D intermediate, we do that in xASL_wrp_Resample.

Or are you saying that we have to save PWI3D again here, because we have potentially merged PWI4D or changed it or anything in between xASL_wrp_Resample and xASL_wrp_Quantify?

Then I agree, but then still it would be nice to have a clean management/wrapper and a single quantification function.

We can continue this conversation in #1572 indeed.

So you can resolve.

jan-petr replied:
Yes, we need new PWI3D. We had PWI4D, PWI3D and PWI calculated and saved previously. But then we do despiking and potentially exclude some volumes. So we need to redo PWI3D based on the PWI4D cause some volumes might be removed. So we need saving again. 

Just explaining - you can resolve.

HenkMutsaerts commented on Functions/Quantification/xASL_quant_ASL.m:
```diff
@@ -105,10 +105,17 @@
 M0_im = double(M0_im);
 
+% Create path to PWI3D based on the PWI4D path (Because PWI4D_Path can be a path to a merged NII)
+PWI3D_Path = PWI4D_Path;
```
I would first take the file and change it from there, which seems a bit more secure to me:

[fPath, fFile] = xASL_fileparts(PWI4D_Path);
if ~strcmp(fFile(1:5), 'PWI4D')
      error('Incorrect PWI4D NIfTI path');
elseif length(fFile)==5
      fFile =[];
else
      fFile = fFile(6:end);
end
PWI3D_Path = fulllfile(fPath, ['PWI3D' fFile '.nii']);

jan-petr replied:
Yes. Sure, it's slightly more secure, so doing that. You can resolve.

jan-petr replied:
But it can be simplified and will work
[fPath, fFile, fExt] = xASL_fileparts(PWI4D_Path);
if ~strcmp(fFile(1:5), 'PWI4D')
	error('Incorrect PWI4D NIfTI path');
else
	PWI3D_Path = fulllfile(fPath, ['PWI3D' fFile(6:end) fExt]);
end

It will work also for PWI4D.nii.

HenkMutsaerts reviewed (changes requested):
Nearly done

HenkMutsaerts committed:
#1543 xASL_wrp_Quantify: minor masking comment

jan-petr committed:
#1543 Minor fixes

jan-petr committed:
#1543 Revamp quant_FSL

jan-petr committed:
#1543 Basic revamp of xASL_sub_FSLOptions

jan-petr committed:
#1543 Fix quant_FSL for PASL

jan-petr committed:
#1543 Minor bugs

jan-petr committed:
#1543 Finalize quant_FSL

jan-petr committed:
#1543 Minor fix

jan-petr committed:
#1543 Fix quant_FSL

BeatrizPadrela committed:
#1543 M0 scaling division when merging PWI4D

BeatrizPadrela committed:
#1543 M0 scaling division when merging PWI4D - in dataPar instead of studyPar

BeatrizPadrela commented:
New name for the scaling variable – discussed with Jan

![image](https://github.com/ExploreASL/ExploreASL/assets/73699072/46dea54d-8475-42dc-94d5-2cdeb7682634)

BeatrizPadrela committed:
#1543 changing the name variable to SessionMergingPWIScaling

HenkMutsaerts committed:
#1543 xASL_io_ASLSubtractionAveraging: add comments

HenkMutsaerts committed:
#1543 xASL_wrp_RegisterASL: fix correct path loading xASL_io_ASLSubtractionAveraging

jan-petr committed:
#1543 wrp_ResampleASL Minor fix

HenkMutsaerts reviewed (changes requested):
Went through all conversations now I believe :)

jan-petr committed:
#1543 wrp_RealignASL Minor fix

jan-petr committed:
#1543 io_ASLSubtractionAveraging Fix reference saving

HenkMutsaerts committed:
#1543 xASL_io_SaveNifti: manage overwriting

jan-petr commented on External/SPMmodified/xASL/xASL_io_SaveNifti.m:
```diff
@@ -1,4 +1,4 @@
+function xASL_io_SaveNifti(pathOrigNifti, pathNewNifti, imNew, nBits, bGZip, changeMat, bCopyOrigJson, JsonFields, bLegacy2BIDS, bOverwrite)
 % Save a file to a Nifti format, while taking the parameters from another file
 %
 % FORMAT: xASL_io_SaveNifti(pathOrigNifti, pathNewNifti, imNew[, nBits, bGZip, changeMat, bCopyOrigJson, JsonFields])
```
Add bOverwrite - just a comment for myself. I'll fix this.

jan-petr replied:
Done in 39d293ee

jan-petr commented on External/SPMmodified/xASL/xASL_io_SaveNifti.m:
```diff
@@ -87,6 +89,10 @@ function xASL_io_SaveNifti(pathOrigNifti, pathNewNifti, imNew, nBits, bGZip, cha
 	bLegacy2BIDS = true;
 end
 
+if nargin < 10 || isempty(bOverwrite)
```
Make sure the length is three. Note for myself, I will fix this.
true -> "true true true"
false -> false false false"
true false -> true false true
false false -> false false true

HenkMutsaerts replied:
Sure though with length<3 I would also issue a warning, since they may not have thought about skipping JSON/MAT

jan-petr replied:
Good point. if length == 1, then we prolong and issue a warning. If length == 2, then we throw an error because it's actually a bit unclear if [1 0] should extend to [1 0 1] or [1 0 0] etc., so best to throw an error.

Done in: 39d293ee

jan-petr commented on External/SPMmodified/xASL/xASL_io_SaveNifti.m:
```diff
@@ -229,7 +235,16 @@ function xASL_io_SaveNifti(pathOrigNifti, pathNewNifti, imNew, nBits, bGZip, cha
+    elseif ~bOverwrite(3)
+        % don't do anything
+    else
+        warning('Pre-existing -mat-file')
```
@HenkMutsaerts This option will never be reached. As bOverwrite wil be either true or false, but never this third option. If you agree, I'll remove this.

HenkMutsaerts replied:
Is that true? What about when bOverwrite is a NaN? If we catch those options at the start I'm fine removing this here.

I thought of a way to have a warning only if the user specified something, that doesn't work. But I agree with you we should remove it here and move it to the start

jan-petr replied:
Agree: If NaN or Inf is detected, we throw an error. Done in 39d293ee

Of course if user didn't specified anything, then the default values kick in :)

HenkMutsaerts replied:
Why not simply adding an "else"? You never know that you catch all cases...

But I'm fine otherwise, you can resolve this conversation.

jan-petr replied:
The "else" branch contained only "%Do nothing", so no point adding an "else" that has a blank statement.

jan-petr commented on External/SPMmodified/xASL/xASL_io_SaveNifti.m:
```diff
@@ -259,41 +274,45 @@ function xASL_io_SaveNifti(pathOrigNifti, pathNewNifti, imNew, nBits, bGZip, cha
+        xASL_delete(pathNewNifti);
+    elseif ~bOverwrite(1)
+        % don't do anything
+    else
```
Same as above.

jan-petr commented on External/SPMmodified/xASL/xASL_io_SaveNifti.m:
```diff
@@ -259,41 +274,45 @@ function xASL_io_SaveNifti(pathOrigNifti, pathNewNifti, imNew, nBits, bGZip, cha
+        xASL_delete(pathNewJson);
+    elseif ~bOverwrite(2)
+        % don't do anything
+    else
```
Same as above.

jan-petr commented on Modules/Module_Processing/Module_ASL/xASL_wrp_RealignASL.m:
```diff
@@ -526,7 +526,7 @@ function xASL_wrp_RealignASL(x, bASL)
 		end
 
-        xASL_io_SaveNifti(x.P.Path_ASL4D, x.P.Path_despiked_ASL4D, NewIm, 32, 0, [], 1, jsonFields, true);
+        xASL_io_SaveNifti(x.P.Path_ASL4D, x.P.Path_despiked_ASL4D, NewIm, 32, 0, [], 1, jsonFields, true, [true true, false]);
```
VERY GOOD POINT!! Shall I check for all other newly added SaveNiftis? Maybe we have messed up this on more occassions.

HenkMutsaerts replied:
We can check this, definitely. I guess that we do this correctly on other locations, because you always want to have the JSON & .mat of the new NIfTI. Here specifically we recreate the .mat ourselves.

One other way to do this, is that if we recreate JSON or .mat "manually" (i.e., outside of xASL_io_SaveNifti), it is safer to do this after xASL_io_SaveNifti rather than before.

jan-petr replied:
OK. I guess we can resolve then

jan-petr reviewed (commented):
See my comments.

HenkMutsaerts committed:
#1543 xASL_* minor improvements

jan-petr committed:
#1543 Minor edits

jan-petr committed:
#1543 io_SaveNifti: Minor edits

jan-petr committed:
#1543 module_ASL Fix SaveCBF4D parameter check

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -429,20 +429,16 @@
-	if isfield(x.modules.asl, 'SaveCBF4D') && x.modules.asl.SaveCBF4D
-        if x.Q.nUniqueInitial_PLD>1 || x.modules.asl.bQuantifyMultiTE
-            warning('Saving CBF4D was requested but not implemented yet for multi-PLD or multi-TE, skipping');
+	if x.modules.asl.SaveCBF4D && x.Q.nUniqueInitial_PLD == 1 && x.modules.asl.bQuantifyMultiTE == false
```
bQuantifyMultiTE should not have influence on computing CBF4D. If we ask CBF4D, we do both CBF and CBF4D. For both single- or multi-PLD, this is possible only when you have multiple volumes.

So instead of this here, I would just issue a warning when SaveCBF4D is asked & we have multi-PLD or multi-TE, stating that this is not implemented yet.

jan-petr replied:
We can only do CBF4D for those sequences where we do CBF quantification inside ExploreASL. When quantifying by BASIL, we can only do CBF but not CBF4D. So multiTE and multi-PLD cannot save CBF4D. 

As discussed - we left only a safety check here. The proper check and warnings are done at line 824 in the main function for checking inputs - as discussed.

HenkMutsaerts replied:
If we do a single-PLD of one of the multi-PLDs, you can do a CBF4D on this.

But I agree with you. I would just have the checks, warnings, etc above, so we can remove these redundant checks here and simplify the code. Otherwise, you can have really a lot of if statements throughout, because of course there many more reasons to skip a part of the code.

Any way, you can resolve :)

jan-petr replied:
Yes, true, a single-PLD can be selected, but that's probably not what we want to avoid confusion.

OK - no prob. The warning is redundant, so no point in keeping it. Done in 696ad616

I'm leaving this FYI. YOu can resolve without checking the code.

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -824,6 +820,14 @@
     end
 end
 
+% Saving of CBF4D is only possible for single TE, single PLD sequences
```
This SaveCBF4D management should happen above where we issue the warning (or the warning goes here, in any case, I would manage SaveCBF4D in this function on 1 location.

jan-petr replied:
We have discussed this and you wanted to move SaveCBF4D management here to this location where we manage all other parameters. So it's all (management and warnings) on a single location. 

The other place contains just a safety catch without warnings. That catch is not necessary and we can remove it, but there's no harm in having more safety catches.

HenkMutsaerts replied:
OK then I would vouch to remove it at the other location. Of course it is safer to have it at more locations, but it reduces code readability. You can either do this or disagree, but in any case you can resolve this conversation :)

jan-petr replied:
Agree - removed in the main function as explained in the comment above.

HenkMutsaerts reviewed (changes requested):
Minor minor

jan-petr committed:
#1543 Save Path_PWI4D_used

jan-petr committed:
#1543 Fix PWI scaling during merging

jan-petr committed:
#1543 Remove TODO comments

jan-petr committed:
#1543 module_ASL Minor fixes

HenkMutsaerts reviewed (approved):
This can now finally be merged! Though note that the MergingScaling parameter should not be in here. Let's discuss this separately.

My point is that you can easily make a vector from the factors that have an influence on the scaling, such as FA. And then these merge properly such as TE, PLD, etc. Then you take this into account for quantification (doing the scaling volume-wise, before subtraction/averaging). Because this will be more fool-proof (more difficult for the users to forget) and more reproducible. Unless this is very difficult or a lot of work. In any case, I'm fine moving this to the future; but note that you will confuse users because they first need a scaling parameter which would be removed soon.

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -755,18 +755,47 @@
+% Read and check a corresponding list with merging scalings
+if ~isfield(x.modules.asl, 'SessionMergingScaling')
+	% By default set as empty
+	x.modules.asl.SessionMergingScaling = {};
```
Also comment here that it is a list of list, I thought it should be a [] but you are right, which I now only find out 2 lines further

jan-petr replied:
Good point. Added a comment.

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -755,18 +755,47 @@
+	end
+end
+
+% Compare the two lists and report an error if the lengths do not correspond
```
Note how much easier it is to have a FA vector instead of this quick and dirty manual parameter that can go wrong on many levels ;)

jan-petr replied:
We haven't figured out how to make it work with FA alone. Otherwise, this would have of course been the number 1 solution.

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -755,18 +755,47 @@
 % e.g., if we want to concatenate ASL_1 & ASL_2, this will only contain both sessions for the xASL_module_ASL iteration of ASL_2
 x.modules.asl.sessionsToMerge = {};
-
+x.modules.asl.sessionsToMergeScaling = [];
```
Why is this [] and for sessionsToMerge {}?
Would make this equal

jan-petr replied:
sessionsToMerge is a list of strings, so it has to be a cell.
sessionsToMergeScaling is a list of number = a normal array, so it's logical to make it a normal vector. We could make it a cell, but that would be really weird.
Note that this is the current list for a given session, so not a list of lists anymore.

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_VisualQC_ASL.m:
```diff
@@ -134,12 +134,13 @@ function xASL_wrp_VisualQC_ASL(x)
+        if ~isfield(x.P, 'Path_PWI4D_used') || ~xASL_exist(x.P.Path_PWI4D_used, 'file')
             [ControlIM, LabelIM] = xASL_quant_GetControlLabelOrder(xASL_io_Nifti2Im(x.P.Path_ASL4D));
-            xASL_io_SaveNifti(x.P.Path_ASL4D, x.P.Path_PWI4D, ControlIM-LabelIM);
+			x.P.Path_PWI4D_used = x.P.Path_PWI4D;
```
I would not do this. We have a single management location for choosing which PWI4D path to use, and to perhaps revert to an older one. Such intermediate fixes to avoid the pipeline crashing are nice, but you have just created a potential bug!

Think about the situation that for whatever reason the PWI4D_used definition went wrong. Then, the user will look at the wrong visual QC! I would remove this, and in the previous instance where we define this path, we issue a warning.

Note that this also means that the merging and definition of PWI4D should also happen when the status file for quantification already exists. So these should happen just above the quantification part.

jan-petr replied:
The first two paragraphs - I agree, I merely worked what was already in the code. See the comment below.

Regarding the third part - I don't understand. The ASL module does first Resampling, the merging, then quantification. PWI4D merging has to happen before quantification.

HenkMutsaerts replied:
My point is that if the status files already exist, the definition of PWI4D_merged should still be done.

E.g., situation 1:
ASL module crashes after the quantification, and only the xASL_wrp_VisualQC is redone when restarting the ASL module (then we need to have the definition x.P.PWI4D_used redefined, even though it will skip the quantification because that status file already exists) -> so we move the PWI4D_used definition to before the status check for quantification

jan-petr replied:
Good point. Done.

HenkMutsaerts commented on Modules/Module_Processing/Module_ASL/xASL_wrp_VisualQC_ASL.m:
```diff
@@ -134,12 +134,13 @@ function xASL_wrp_VisualQC_ASL(x)
         % If segmented files exist, then the PVgm should already be prepared previously
 
-        if ~xASL_exist(x.P.Path_PWI4D, 'file')
+        if ~isfield(x.P, 'Path_PWI4D_used') || ~xASL_exist(x.P.Path_PWI4D_used, 'file')
```
Thinking about it, this whole if statement is suboptimal, because you are catching something that should have happened before, and trying to decide in a detailed code part what we should have done before.

So: we should prefer a pipeline crash over sacrificing reliability

jan-petr replied:
Good point. I was merely updating the code, not inspecting it further. But you are right that fixing this might create confusion, while in practice, the file should be always there unless things went wrong.

So I'm deleting this whole update. Done

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -415,6 +415,8 @@
 		x.P.Path_PWI4D_used = x.P.Path_PWI4D;
 		x.P.Pop_Path_PWI4D_used = x.P.Pop_Path_PWI4D;
 	end
+	% Note: The paths Path_PWI4D_used can differ from a simple subtraction as it can point out to version that has some outlying volumes excluded.
```
Change this note to:

% Note: The paths x.P.(Pop_)Path_PWI4D_used can differ from a PWI4D obtained from a simple subtraction as we may have merged sessions, and the user or ExploreASL may have removed volumes from PWI4D

jan-petr replied:
Agree. Done.

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -415,6 +415,8 @@
 		x.P.Pop_Path_PWI4D_used = x.P.Pop_Path_PWI4D;
 	end
+	% Note: The paths Path_PWI4D_used can differ from a simple subtraction as it can point out to version that has some outlying volumes excluded.
+	% That's why we save this actual path used so that the correct version can be reused for xASL_wrp_VisualQC
```
Change this note to:

% That's why we save and use x.P.(Pop_)Path_PWI4D_used in memory, such that the correct NIfTI will be used later in the pipeline (e.g., in xASL_wrp_VisualQC)

jan-petr replied:
Agree. Replaced

HenkMutsaerts reviewed (changes requested):
Minor points

jan-petr committed:
#1543 Minor fixes

jan-petr commented on Modules/xASL_module_ASL.m:
```diff
@@ -760,18 +758,47 @@
 for iList=1:nLists
 	if ~isempty(x.modules.asl.SessionMergingList{iList}) && sum(ismember(x.SESSION, x.modules.asl.SessionMergingList{iList}))
-		currentSortedList = sort(x.modules.asl.SessionMergingList{iList});
+		[currentSortedList, indexSorted] = sort(x.modules.asl.SessionMergingList{iList});
```
Make sure that a list [1 1.27] is correctly converted to a list of lists.

jan-petr reviewed (commented):
One small stuff

HenkMutsaerts reviewed (changes requested):
Looks all good. One conversation still open, and we need to discuss how to solve the scaling nicely

jan-petr committed:
#1543 module_ASL: Run merging when module not finished

jan-petr committed:
#1543 module_ASL: Fix Scaling parameter parsing

HenkMutsaerts committed:
#1543 xASL_module_ASL: improve PWI4D_used management

HenkMutsaerts reviewed (changes requested):
I have added a commit that I believe helps, but check if you agree.

1. Why is xASL_im_MergePWI4D a separate function, and is all the other merging defined in a subfunction? (I don't mind much, just asking)
2. You added this statement for merging:
x.mutex.HasState(StateName{iState-4}) && (~x.mutex.HasState(StateName{iState}) || bOutput)

which I don't understand. Why do you only want to define path_PWI4D_used when you have finished resampling && not yet quantification or you want output?
A) it is IMHO too little overhead to add an if-statement (just add an if exist and we always run the same code
B) how does bOutput make sense, we never need PWI4D_used as output?

jan-petr commented:
> 1. Why is xASL_im_MergePWI4D a separate function, and is all the other merging defined in a subfunction? (I don't mind much, just asking)

It's a separate function that handles the merging of parameters and images. We have agreed upon that. There are no important subfunctions in it. Is that good as an answer to your question?

HENK: No I know that :) My question is why this is a separate function and the merging admin (variables definition etc) is a subfunction. It seems arbitrary why one thing is a separate function and the other is not. Not very important.

JAN: I don't see which subfunction you mean. But lets leave it if it is unimportant.

> 2. You added this statement for merging:
>    x.mutex.HasState(StateName{iState-4}) && (~x.mutex.HasState(StateName{iState}) || bOutput)
> 
> which I don't understand. Why do you only want to define path_PWI4D_used when you have finished resampling && not yet quantification or you want output? A) it is IMHO too little overhead to add an if-statement (just add an if exist and we always run the same code B) how does bOutput make sense, we never need PWI4D_used as output?

It had a very good reason to do it like this:
1) In case we finished ASL module completely in the first run, we of course don't need this extra IF. 
2) However, if we run the ASL module again (e.g. adding an extra subject), it will do the merging again for all subjects. Which is an extra overhead. And what is worse - it could mean (if dataPar.json was modified but the quantification isn't run again) that the merged PWI and CBF are done on different images. So that's why I added the mutex.
3) But in case quantification was done, but QC not, you wanna make sure that the PWI_used is initialized, that's why I added bOutput, which is false when the module is fully completed, but true otherwise.

HENK: Is the merging a lot of overhead? It only reads a json, that is not a lot of overhead right?

JAN: It's a cosiderable overhead when you have 100+ subjects. It's definitively 1000 times larger overhead to I/O NII instead of keeping a single IF :)

HENK: In the past we had something that would return xASL_module_ASL when it had 999_ready, not sure why we removed this.
JAN: We still have 999_ready. But 999_ready doesn't skip the entire module, but still runs through with bOutput turned OFF. So that's why I used bOutput as a proxy of 999_ready. We can of course do it some other way, but some kind of a check is needed.

So the if statement had a very good reason to avoid unnecessary calculations, reduce overhead, avoid inconsistencies.
I agree that we could check for QC-mutex instead of bOutput instead if you want. We also need to rewrite the code accordingly or revert your commit.

HENK: Probably checking that quantification OR QC mutex doesn't exist is then the best indeed. You can keep my "improved comments" and move this improved mutex checking back?
JAN: OK, so I'll move in all the mutexes, but instead of bOutput, I'll put in a QC-mutex. See my fix committed.

jan-petr committed:
#1543 module_ASL: Minor fix

HenkMutsaerts commented on Modules/xASL_module_ASL.m:
```diff
@@ -407,14 +407,16 @@
-	% Note: The paths x.P.(Pop_)Path_PWI4D_used can differ from a PWI4D obtained from a simple subtraction as we may have merged sessions, 
-	% and the user or ExploreASL may have removed volumes from PWI4D. That's why we save and use x.P.(Pop_)Path_PWI4D_used in memory, such 
-	% that the correct NIfTI will be used later in the pipeline (e.g., in xASL_wrp_VisualQC)
+if (x.mutex.HasState(StateName{iState-4}) && (~x.mutex.HasState(StateName{iState}))) ||...
```
Simplification:

You can remove (x.mutex.HasState(StateName{iState-4}) because 1) without resampling done, it wouldn't get here (it doesn't proceed to mutex+1 if mutex hasn't been completed), so we could write this for each mutex. Plus, this 2) becomes buggy if we would change the order of mutexes. So I would keep this simple. Plus 3) if this were to be the case, we actually want a crash! We rather want a crash than wrong results... (consider that a user could get results without taking the merged PWI4D but unmerged one)

We hardly use WADQC, so we could also move WADQC as step 5 of visualQC to make this code easier; we don't have any other dependencies of the mutex stats being n=10 right?

Actually I now already did this, let me know if you disagree :)

HenkMutsaerts committed:
#1543 Move WAD-QC -> xASL_wrp_VisualQC_ASL
